### PR TITLE
improve export

### LINKED ISSUE #490: Suggestion: Have note files actually be mark down files
Would be nice if the files could be viewed by any program that can view markdown.  One suggestion is to put your internal data you need as a comment at the top of the file:

```
<!--
type: "MARKDOWN_NOTE"
folder: "02d0f982bd56ad197fe5"
title: "My Title"
tags: []
isStarred: false
createdAt: "2017-04-23T03:30:44.232Z"
updatedAt: "2017-04-23T03:32:00.174Z"
-->

My Note Content
```

### PR DESCRIPTION
**Pull request by daiyam**
## Description

This change improves the export in several areas:
- you can add YAML front matter to export markdown files
- the codes (to export a note, a folder, a storage, in Markdown or in HTML) are, now, an unique code
- you can export a folder or a storage in HTML files
- the attachement folder can be unique for each note
- better labels in context menus
- show a message when the export of a folder or a storage, is done
- export dependencies to generate diagrams and style code blocks

The preferences:
screenshot

A context menu:
screenshot

screenshot

## Issue fixed

- #490 
- #2675
- #2648
- #3305

## Type of changes

- :radio_button: Bug fix (Change that fixed an issue)
- :white_circle: Breaking change (Change that can cause existing functionality to change)
- :radio_button: Improvement (Change that improves the code. Maybe performance or development improvement)
- :white_circle: Feature (Change that adds new functionality)
- :white_circle: Documentation change (Change that modifies documentation. Maybe typo fixes)

## Checklist:

- :radio_button: My code follows [the project code style](docs/code_style.md)
- :white_circle: I have written test for my code and it has been tested
- :radio_button: All existing tests have been passed
- :radio_button: I have attached a screenshot/video to visualize my change if possible

### CODE REVIEW

**Comment by ZeroX-DG** [comment]
I’m thinking of a different name for the export tab. Maybe Markdown tab the export section in it? We may need to add some markdown related configurations in the future. Another reason is dedicating a whole tab just for export is.....space consuming.

**Comment by daiyam** [comment]
No reason but the `Prefix attachment folder` is not a Markdown thing.

File: browser/main/lib/dataApi/attachmentManagement.js [hunk 1]
```diff
@@ -386,6 +386,17 @@ function removeStorageAndNoteReferences (input, noteKey) {
...
 }
 
+/**
+ * @description replace all :storage references with given destination folder.
+ * @param input Input in which the references should be deleted
+ * @param noteKey Key of the current note
+ * @param destinationFolder Destination folder of the attachements
+ * @returns {String} Input without the references
+ */
+function replaceStorageReferences (input, noteKey, destinationFolder) {
```
Comment (ehhc) [inline browser/main/lib/dataApi/attachmentManagement.js]: Would it be possible that you add some tests for this method? e.g. for a note with one attachment and one test with multiple attachments?

File: browser/main/lib/dataApi/attachmentManagement.js [hunk 2]
```diff
@@ -386,6 +386,17 @@ function removeStorageAndNoteReferences (input, noteKey) {
   return input.replace(new RegExp(mdurl.encode(path.sep), 'g'), path.sep).replace(new RegExp(STORAGE_FOLDER_PLACEHOLDER + '(' + escapeStringRegexp(path.sep) + noteKey + ')?', 'g'), DESTINATION_FOLDER)
 }
 
+/**
+ * @description replace all :storage references with given destination folder.
+ * @param input Input in which the references should be deleted
```
Comment (ehhc) [inline browser/main/lib/dataApi/attachmentManagement.js]:
```suggestion
 * @param input Input in which the references should be replaced
```

**Comment by ZeroX-DG** [comment]
@daiyam can you resolve the conflict? I'll review this soon :+1:

**Comment by daiyam** [comment]
@ZeroX-DG I'm on it but I'm sure there will be conflict with #3146... and I have to fix #3554

**Comment by ZeroX-DG** [comment]
@daiyam yea, it might conflict with the name:) I'll wait until you fix your babies https://github.com/BoostIO/Boostnote/issues/3554 :smile:

**Comment by daiyam** [comment]
@ZeroX-DG here, you are.

**Comment by daiyam** [comment]
@ZeroX-DG I should be able to fix the PR on this week-end.

**Comment by daiyam** [comment]
@ZeroX-DG I found some time to fix the PR...
In the regrouped function, I've added `<` and `\s` as possible terminators used when the reference is inside a data block (ex: gallery)

**Comment by ZeroX-DG** [comment]
@daiyam looks good to me. Except you didn't remove the deletion of `isStacking` variable. I'll test this more when I get home.

**Comment by daiyam** [comment]
Oops. I will check that ;)

**Comment by daiyam** [comment]
@ZeroX-DG I was trying to understand what was this flag... until read a7ead67c2d91a6f0e3297688d619eb1b62d39ff9

**Comment by ZeroX-DG** [comment]
Ah yes, I should have told you what it does...sorry about that :smile:

File: browser/main/Detail/MarkdownNoteDetail.js [hunk 3]
```diff
@@ -474,6 +474,7 @@ class MarkdownNoteDetail extends React.Component {
           value={note.content}
           storageKey={note.storage}
           noteKey={note.key}
+          isStacking={isStacking}
```
Comment (ZeroX-DG) [inline browser/main/Detail/MarkdownNoteDetail.js]: I think you forgot to extract `isStacking` from `this.state` first. Right now it's undefined:) Also can you do the same thing to `MarkdownEditor` above this code?

Comment (daiyam) [reply browser/main/Detail/MarkdownNoteDetail.js]: done 😌

**Pull request merged**